File: sdks/node/src/plugins/droid.ts
import type { AgentPlugin, InstallConfig, PluginStatus } from './base.js';
import {
  AGENT_KEY,
  buildGroup,
  runtimeContracts,
  type DroidGroup,
  type ToolEvent,
} from './droid-contract.js';
import {
  activeDocument,
  additionsForTarget,
  effectiveHooks,
  hookBlock,
  installationDocuments,
  renderDocument,
  sourceDocuments,
  type DroidSources,
  type RenderedDocument,
} from './droid-config.js';
import {
  commitDroidInstallation,
  commitDroidUninstall,
  preflightDroidInstallation,
  prepareDroidInstallation,
  prepareDroidUninstall,
} from './droid-installation.js';
import { displayConfigPath, readSources, runtimeFilesExist } from './droid-io.js';
import { SUPPORTED_AGENTS } from './registry.js';

const entry = SUPPORTED_AGENTS.get(AGENT_KEY)!;

function renderInstallation(
  sources: DroidSources,
  guardPath: string,
  auditPath: string,
): RenderedDocument[] {
  const target = activeDocument(sources);
  const groups = new Map<ToolEvent, DroidGroup>([
    ['PreToolUse', buildGroup(guardPath)],
    ['PostToolUse', buildGroup(auditPath)],
  ]);
  return installationDocuments(sources).map((document) => renderDocument(
    document,
    undefined,
    additionsForTarget(document, target, groups),
  ));
}

function renderUninstall(sources: DroidSources, agentId?: string): RenderedDocument[] {
  return sourceDocuments(sources).map(
    (document) => renderDocument(document, agentId, new Map()),
  );
}

export const droidPlugin: AgentPlugin = {
  managesRuntime: true,

  async preflightInstall(config: InstallConfig): Promise<void> {
    const sources = await readSources();
    await preflightDroidInstallation(config, sources);
  },

  async install(config: InstallConfig): Promise<void> {
    const sources = await readSources();
    const paths = await preflightDroidInstallation(config, sources);
    const rendered = renderInstallation(sources, paths.guardPath, paths.auditPath);
    const prepared = await prepareDroidInstallation(config, sources, rendered);
    await commitDroidInstallation(prepared);
    console.log(`  Factory Droid hooks: ${activeDocument(sources).filePath}`);
    console.log('  Factory Droid: run /hooks to review the Elydora hook changes.');
  },

  async uninstall(agentId?: string): Promise<void> {
    const sources = await readSources();
    const prepared = await prepareDroidUninstall(renderUninstall(sources, agentId));
    await commitDroidUninstall(prepared);
  },

  async status(): Promise<PluginStatus> {
    const sources = await readSources();
    const contracts = runtimeContracts(effectiveHooks(sources));
    const hookConfigured = hookBlock(sources) === undefined && contracts.length > 0;
    const hookScriptExists = hookConfigured && await runtimeFilesExist(contracts);
    return {
      installed: hookConfigured && hookScriptExists,
      agentName: AGENT_KEY,
      displayName: entry.name,
      hookConfigured,
      hookScriptExists,
      configPath: displayConfigPath(sources),
    };
  },
};

File: sdks/go/cmd/elydora/plugins/droid.go
package plugins

import "fmt"

// DroidPlugin manages Factory Droid native user hooks.
type DroidPlugin struct {
	rename renameFunc
}

// ManagesGuardRuntime reports that Droid commits its guard, audit runtime,
// credentials, and hook documents in one transaction.
func (p *DroidPlugin) ManagesGuardRuntime() bool {
	return true
}

// PreflightInstall validates every hook source, policy layer, credential,
// matcher, and runtime identity before any write.
func (p *DroidPlugin) PreflightInstall(config InstallConfig) error {
	sources, err := readDroidSources()
	if err != nil {
		return err
	}
	_, _, err = preflightDroidInstallation(config, sources)
	return err
}

func renderDroidInstallation(
	sources *droidSources,
	guardPath, auditPath, nodePath, runtimeRoot string,
) ([]*droidRenderedDocument, error) {
	target := activeDroidDocument(sources)
	groups := map[string]map[string]any{
		"PreToolUse":  buildDroidGroup(nodePath, guardPath),
		"PostToolUse": buildDroidGroup(nodePath, auditPath),
	}
	documents := droidInstallationDocuments(sources)
	rendered := make([]*droidRenderedDocument, 0, len(documents))
	for _, document := range documents {
		item, err := renderDroidDocument(
			document,
			"",
			runtimeRoot,
			droidAdditionsFor(document, target, groups),
		)
		if err != nil {
			return nil, err
		}
		rendered = append(rendered, item)
	}
	return rendered, nil
}

func (p *DroidPlugin) Install(config InstallConfig) error {
	sources, err := readDroidSources()
	if err != nil {
		return err
	}
	paths, nodePath, err := preflightDroidInstallation(config, sources)
	if err != nil {
		return err
	}
	rendered, err := renderDroidInstallation(
		sources,
		paths.guardPath,
		paths.auditPath,
		nodePath,
		paths.runtimeRoot,
	)
	if err != nil {
		return err
	}
	prepared, err := prepareDroidInstallation(config, sources, rendered)
	if err != nil {
		return err
	}
	if err := commitDroidInstallation(prepared, p.rename); err != nil {
		return err
	}
	fmt.Printf("Factory Droid hooks: %s\n", activeDroidDocument(sources).filePath)
	fmt.Println("Factory Droid: run /hooks to review the Elydora hook changes.")
	return nil
}

func renderDroidUninstall(
	sources *droidSources,
	agentID, runtimeRoot string,
) ([]*droidRenderedDocument, error) {
	documents := droidSourceDocuments(sources)
	rendered := make([]*droidRenderedDocument, 0, len(documents))
	for _, document := range documents {
		item, err := renderDroidDocument(
			document,
			agentID,
			runtimeRoot,
			map[string]map[string]any{},
		)
		if err != nil {
			return nil, err
		}
		rendered = append(rendered, item)
	}
	return rendered, nil
}

func (p *DroidPlugin) Uninstall(agentID string) error {
	sources, err := readDroidSources()
	if err != nil {
		return err
	}
	runtimeRoot, err := droidRuntimeRoot()
	if err != nil {
		return err
	}
	rendered, err := renderDroidUninstall(sources, agentID, runtimeRoot)
	if err != nil {
		return err
	}
	prepared, err := prepareDroidUninstall(rendered)
	if err != nil {
		return err
	}
	return commitDroidUninstall(prepared, p.rename)
}

func (p *DroidPlugin) Status() (PluginStatus, error) {
	entry := SupportedAgents[droidAgentKey]
	status := PluginStatus{AgentName: droidAgentKey, DisplayName: entry.Name}
	sources, err := readDroidSources()
	if err != nil {
		return status, err
	}
	status.ConfigPath = displayDroidConfigPath(sources)
	if droidHookBlocked(sources) != nil {
		return status, nil
	}
	runtimeRoot, err := droidRuntimeRoot()
	if err != nil {
		return status, err
	}
	contracts := droidRuntimeContracts(effectiveDroidHooks(sources), runtimeRoot)
	status.HookConfigured = len(contracts) > 0
	if !status.HookConfigured {
		return status, nil
	}
	status.HookScriptExists, err = droidRuntimeFilesExist(contracts)
	if err != nil {
		return status, err
	}
	status.Installed = status.HookScriptExists
	return status, nil
}
